### PR TITLE
Implement support for the free-threaded build of CPython 3.13

### PR DESCRIPTION
**Pull request by lysnikolaou**
## What do these changes do?

Implement support for the free-threaded build of CPython 3.13!

In more detail, this PR does the following:

- Add a CI job that runs the tests under the free-threaded build.
- Enable building free-threaded wheels.
- Add `pythoncapi_compat.h` and move away from all the borrowed reference APIs.
- Mark extension module as thread-safe
- Add locking around `pair_list_global_version`.

## Are there changes in behavior for the user?

There shouldn't be any.

## Related issue number

Related to #1014.

## Checklist

- [x] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes

### CODE REVIEW

**Comment by bdraco** [comment]
The changes look good to me and I didn't see any ref count issues, however I am by no means an expert on the C API

**Review by bdraco — APPROVED** [review]

**Comment by bdraco** [comment]
Assuming the CI passes (might take a while to run since there is a lot of activity right now), I'd like another maintainer who has more experience with the C API to take a look before merging.

**Comment by bdraco** [comment]
Thanks @lysnikolaou

**Comment by bdraco** [comment]
It looks like one of the existing actions unrelated to this PR is using an old upload-artifact.  It would be nice if github told us which one.

**Comment by lysnikolaou** [comment]
Yeah, I saw that as well. Opened #1016 for it.

**Comment by bdraco** [comment]
Looks like its almost there 👍

**Comment by lysnikolaou** [comment]
Looks like it's good to go on my side!

**Review by bdraco — APPROVED** [review]

**Comment by bdraco** [comment]
LGTM, but I'd still like another set of 👀 https://github.com/aio-libs/multidict/pull/1015#issuecomment-2358761183 before merging

**Comment by lysnikolaou** [comment]
Friendly ping here!

File: CHANGES/1015.feature.rst
Comment (webknjaz) [inline CHANGES/1015.feature.rst:1]: I think we need another change note of the `packaging` type saying
```rst
Started publishing wheels made for the free-threaded build of CPython 3.13 -- by :user:`lysnikolaou`.
```

@bdraco WDYT?

File: .github/workflows/ci-cd.yml
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:168]:
```suggestion
          pyver: 3.13-freethreading  # this is still tested within cibuildwheel
```

File: .github/workflows/ci-cd.yml
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:170]:
```suggestion
          pyver: 3.13-freethreading  # this is still tested within cibuildwheel
```

File: .github/workflows/ci-cd.yml
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:153]: let's follow the existing convention of not adding unnecessary quoting
```suggestion
        - 3.13-freethreading
```

File: .github/workflows/ci-cd.yml
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:203]:
```suggestion
      if: >-
        !endsWith(matrix.pyver, '-freethreading')
```

**Review by webknjaz — COMMENTED** [review]
Here's my CI/CD feedback. I can't say I'm experienced with C-API, however. Still, I think we can merge this and start a release process to see if more rare architectures also succeed building the new wheels. It should be enough to see if cibuildwheel passes in those jobs that only run on release, since it also runs some testing. We can then abort the release once we know. Or we could just attempt a regular release.

File: CHANGES/1015.feature.rst
Comment (bdraco) [reply CHANGES/1015.feature.rst:1]: LGTM

File: multidict/_multilib/pair_list.h
Comment (asvetlov) [inline multidict/_multilib/pair_list.h:64]: I think that the mutex is not needed here.

Straight loose increment is enough, the version number is an auxiliary thing.
If the global counter slightly desynchronized between threads for short time period -- it is ok, no existing code is broken.

File: .github/workflows/ci-cd.yml
Comment (lysnikolaou) [reply .github/workflows/ci-cd.yml:211]: Actually, the `python-install` id is not used anywhere anymore, so I assume that a new version of `codecov` does not need this. I think we can remove these.

File: multidict/_multilib/pair_list.h
Comment (lysnikolaou) [reply multidict/_multilib/pair_list.h]: Mutex removed. Thanks for the help here!

**Comment by lysnikolaou** [comment]
Thanks for the reviews everyone! I've addressed all of the feedback points.

**Comment by bdraco** [comment]
It looks like we don't have any benchmarks for this repository so can't check the performance impact. I guess I'll need to write some.

File: .github/workflows/ci-cd.yml
Comment (lysnikolaou) [reply .github/workflows/ci-cd.yml]: Oh got it. Unfortunately, `deadsnakes/action` does not have a `python-version` output. Would it make sense to hardcode `3.13` for now?

Comment (webknjaz) [reply .github/workflows/ci-cd.yml]: We could have a new step like
```yaml
- name: Compute runtime Python version
  id: python-install
  run: |
    from os import environ
    from pathlib import Path
    from sys import version_info

    FILE_APPEND_MODE = 'a'

    python_version_str = ".".join(map(str, sys.version_info[:3]))

    with Path(environ['GITHUB_OUTPUT']).open(
            mode=FILE_APPEND_MODE,
    ) as outputs_file:
        print(f'python-version={python_version_str}', file=outputs_file)
```
that would work universally.

File: .github/workflows/ci-cd.yml
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:214]:
```suggestion
        nogil: true
    - name: Compute runtime Python version
      id: python-install
      run: |
        from os import environ
        from pathlib import Path
        from sys import version_info

        FILE_APPEND_MODE = 'a'

        python_version_str = ".".join(map(str, sys.version_info[:3]))

        with Path(environ['GITHUB_OUTPUT']).open(
                mode=FILE_APPEND_MODE,
        ) as outputs_file:
            print(f'python-version={python_version_str}', file=outputs_file)
      shell: python
```

Comment (webknjaz) [reply .github/workflows/ci-cd.yml:214]: Although, we should probably extend that with a free-threading flag or something.

File: .github/workflows/ci-cd.yml
Comment (webknjaz) [reply .github/workflows/ci-cd.yml]: Committed that. Marking this thread as resolved.

**Review by webknjaz — APPROVED** [review]

**Comment by bdraco** [comment]
It looks like the testing fails because there is no cffi wheel for free-threaded 3.13 so it tries to build it and since libffi isn't installed in the cibuildwheel env, it fails

**Review by bdraco — APPROVED** [review]

**Pull request merged**
